addons/binding/org.openhab.binding.homie/src/main/java/org/openhab/binding/homie/internal/MqttConnection.java

```diff
@@ -0,0 +1,142 @@
+        return new MqttConnection(config.getBrokerUrl(), config.getBaseTopic(), MQTT_CLIENTID + "-" + consumer);
+    }
+
+    private void connect() {
+        try {
+            logger.debug("Homie MQTT Connection start");
+            MqttConnectOptions opts = new MqttConnectOptions();
+            opts.setAutomaticReconnect(true);
+            opts.setCleanSession(true);
+            client = new MqttClient(brokerURL, MQTT_CLIENTID + "-" + qualifier, new MemoryPersistence());
+            client.connect(opts);
+            logger.debug("Homie MQTT Connection connected");
+        } catch (MqttException e) {
+            logger.error("MQTT Connect failed", e);
+        }
+    }
+
+    public void subscribe(Thing thing, IMqttMessageListener messageListener) throws MqttException {
+        String topic = String.format("%s/%s/#", basetopic, thing.getUID().getId());
+        resubscribe(topic, messageListener);
+
+    }
+
+    public void listenForDeviceIds(IMqttMessageListener messageListener) {
+        logger.debug("Listening for devices on topic {}", listenDeviceTopic);
+        try {
+            resubscribe(listenDeviceTopic, messageListener);
+        } catch (MqttException e) {
+            logger.error("Failed to subscribe to topic {}", listenDeviceTopic, e);
+        }
+    }
+
+    public void unsubscribeListenForDeviceIds() {
+        try {
+            client.unsubscribe(listenDeviceTopic);
+        } catch (MqttException e) {
+            logger.error("Failed to unsubscribe from topic {}", listenDeviceTopic, e);
+        }
+
+    }
+
+    /**
+     * Subscribe to a channel of a device
+     *
+     * @param channelUID
+     * @param handler
+     */
+    public void subscribeChannel(Channel channel, HomieDeviceHandler handler) {
+
+        String topic = String.format("%s/%s/%s", basetopic, handler.getThing().getUID().getId(),
+                channel.getProperties().get(CHANNELPROPERTY_TOPICSUFFIX));
+        logger.debug("(Re-)Subscribing to topic '{}' to listen for events of channel '{}'", topic, channel.getUID());
+        try {
+            resubscribe(topic, handler);
+            logger.debug("Subscribed to topic {}", topic);
+        } catch (MqttException e) {
+            logger.error("Error (re)subscribing to channel. topic is {}", topic, e);
+        }
+
+    }
+
+    private void resubscribe(String topic, IMqttMessageListener listener) throws MqttException {
+        client.unsubscribe(topic);
+        client.subscribe(topic, SUBSCRIBE_QOS, listener);
+    }
+
+    public void listenForNode(String deviceId, String nodeId, IMqttMessageListener listener) throws MqttException {
+        String topic = String.format("%s/%s/%s/", basetopic, deviceId, nodeId);
+        // Subscribe to type and proplist separately, as the notifications have to arrive first at the listener
+        resubscribe(topic + HomieConventions.HOMIE_NODE_TYPE_ANNOUNCEMENT_TOPIC_SUFFIX, listener);
```

Comment:
To me concatenating two strings using `+` is more readable than using `String.format("%s%s",foo, bar)`. That isn't the case if you got >2 strings. Do you agree?